File: packages/python/src/tbd_agents/resources/agents.py
from __future__ import annotations

from typing import Mapping, Any

from tbd_agents.types import (
    Agent,
    AgentCreate,
    AgentExportBundle,
    AgentImportBundle,
    AgentUpdate,
    ImportResult,
)

from .base import BaseResource


class AgentsResource(BaseResource):
    def create(self, payload: AgentCreate | Mapping[str, Any]) -> Agent:
        return self._post("/agents", payload, model=Agent)

    def list(self) -> list[Agent]:
        return self._get_list("/agents", model=Agent)

    def get(self, agent_id: str) -> Agent:
        return self._get(f"/agents/{agent_id}", model=Agent)

    def update(self, agent_id: str, payload: AgentUpdate | Mapping[str, Any]) -> Agent:
        return self._put(f"/agents/{agent_id}", payload, model=Agent)

    def delete(self, agent_id: str) -> None:
        self._delete(f"/agents/{agent_id}")

    def export(self, agent_id: str | None = None) -> AgentExportBundle:
        path = "/agents/export" if agent_id is None else f"/agents/{agent_id}/export"
        return self._get(path, model=AgentExportBundle)

    def import_(self, payload: AgentImportBundle | Mapping[str, Any]) -> ImportResult:
        return self._post("/agents/import", payload, model=ImportResult)

File: packages/flutter/lib/src/resources/agents.dart
/// Agents resource — `/api/agents`.
library;

import 'base.dart';

/// Provides CRUD + import/export operations for agents.
class AgentsResource extends BaseResource {
  const AgentsResource(super.client);

  Future<List<Map<String, dynamic>>> list() async {
    final data = await client.request('GET', 'agents');
    if (data is List) return data.whereType<Map<String, dynamic>>().toList();
    return const [];
  }

  Future<Map<String, dynamic>> get(String id) async {
    final data = await client.request('GET', 'agents/$id');
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> create(Map<String, dynamic> payload) async {
    final data = await client.request('POST', 'agents', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<Map<String, dynamic>> update(
      String id, Map<String, dynamic> payload) async {
    final data = await client.request('PUT', 'agents/$id', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }

  Future<void> delete(String id) => client.request('DELETE', 'agents/$id');

  /// `GET /api/agents/export` or `GET /api/agents/{id}/export`
  Future<Map<String, dynamic>> export({String? agentId}) async {
    final path = agentId == null ? 'agents/export' : 'agents/$agentId/export';
    final data = await client.request('GET', path);
    return (data as Map<String, dynamic>?) ?? {};
  }

  /// `POST /api/agents/import`
  Future<Map<String, dynamic>> import_(Map<String, dynamic> payload) async {
    final data = await client.request('POST', 'agents/import', body: payload);
    return (data as Map<String, dynamic>?) ?? {};
  }
}
